R/getInsertSizeDistFromBam.R

```diff
@@ -0,0 +1,108 @@
+#'
+#' @examples
+#' if (requireNamespace("Rsamtools", quietly = TRUE)) {
+#'     bamf <- system.file("extdata", "getInsertSizeDistFromBam", "atac_mm10.bam",
+#'                         package = "swissknife")
+#'     isize <- getInsertSizeDistFromBam(bamf)
+#'     attr(isize, "ncapped")
+#'     plot(isize, type = "l",
+#'          xlab = "Insert size (bp)", ylab = "Number of fragments")
+#' }
+#'
+#' @importFrom IRanges IRangesList
+#' 
+#' @export
+getInsertSizeDistFromBam <- function(fname,
+                                     regions = NULL,
+                                     nmax = NA_integer_,
+                                     isizemax = 800) {
+    .assertPackagesAvailable(c("Rsamtools"))
+    if (!is.null(regions) && !inherits(regions, "GRanges")) {
+        stop("'regions' must be either NULL or a GRanges object.")
+    }
+    if (!is.na(nmax) && (!is.numeric(nmax) || length(nmax) != 1L || nmax <= 0)) {
+        stop("'nmax' must be either NA or a numeric scalar greater than zero.")
+    }
+    if (!is.numeric(isizemax) || length(isizemax) != 1L || isizemax <= 0) {
+        stop("'isizemax' must be a numeric scalar greater than zero.")
+    }
+    
+    # set scanning parameters
+    if (is.null(regions)) {
+        regions <- IRanges::IRangesList()
+    }
+    params <- Rsamtools::ScanBamParam(
+        flag = Rsamtools::scanBamFlag(isPaired = TRUE,
+                                      isProperPair = TRUE,
+                                      isUnmappedQuery = FALSE,
+                                      hasUnmappedMate = FALSE,
+                                      isFirstMateRead = TRUE,
+                                      isSecondaryAlignment = FALSE,
+                                      isNotPassingQualityControls = NA,
+                                      isDuplicate = FALSE,
+                                      isSupplementaryAlignment = FALSE),
+        what = c("isize"),
+        which = regions)
+    
+    # open bam file(s)
+    bfl <- Rsamtools::BamFileList(fname, yieldSize = min(2e6, nmax, na.rm = TRUE))
+    
+    # read insert sizes
+    bfl <- open(bfl)
+    isizeL <- lapply(bfl, function(x) {
+        itab <- integer(isizemax)
+        attr(itab, "ncapped") <- 0
+        while (length(i1 <- unlist(Rsamtools::scanBam(x, param = params), use.names = FALSE)) > 0) {
+            icap <- i1 > isizemax
+            if (sum(icap) > 0) {
+                attr(itab, "ncapped") <- attr(itab, "ncapped") + sum(icap)
+                i1[icap] <- isizemax
+            }
+            itab <- itab + tabulate(abs(i1), isizemax)
+            if (!is.na(nmax) && sum(itab) >= nmax) {
+                break
+            }
+        }
+        itab
+    })
+    bfl <- close(bfl)
+    ncapped <- sum(unlist(lapply(isizeL, attr, "ncapped")))
+
+    # report on capping
+    if (ncapped > 0) {
+        warning(ncapped, " long insert sizes were capped at isizemax=", isizemax)
+    }
```

Comment:
Would it make sense to have this be a message rather than a warning? I ususally think there is a problem to be aware of if I see a warning, in this case it's not really a problem. Either way is fine though :)

Reply:
I agree - I'll change it to be a `message` instead.